Refactor release workflow (reduce build time)

Zelldon opened the pull request:
## Description

This PR refactors the release workflow. 

It should increase the:
 
  * maintainability
  * performance (reduces the maven job time from ~15-17 minutes to 9 minutes).

### Details 

**Before:**

![2024-04-18_07-21_1](https://github.com/camunda/zeebe/assets/2758593/7ab58e48-bc2c-4355-8e97-033d59033b94)

**After**

![2024-04-18_07-21](https://github.com/camunda/zeebe/assets/2758593/4792363e-5ed6-4eef-9e4e-d124aadecc0c)


#### Split of jobs

With this PR we split the maven release prepare and perform steps into separate jobs, to separate the concerns. The split allows to make more clear what each step does, and
what each parameter is used for.

For example are we able to reduce the parameters to a minimum, and
skip re-building in the preparing step (as the artifacts are not used in a later stage). This reduces the build time of this step from 5 minutes to 30 seconds (factor 10).

For interesting parts and parameters we add some documentation to better understand the what and why.

I followed mostly [the documentation of the release plugin](https://maven.apache.org/maven-release/maven-release-plugin/index.html) to understand what each parameter and goal does. (I had a lot of learnings)

#### Integration tests

Introducing an integration test profile allows us to exclude the respective modules during the release (we need them in the preparation step as we have to change their versions as well). Makes it possible to not build and release unnecessary modules/classes.

#### Go client build

Previously we built the go client and the dist together. The resulting dist was never used, as the go client is packed in the release perform step again. From now on we skip building the Zeebe dist with the go client, this saves us around 1 minute.

## Related issues

related https://github.com/camunda/zeebe/pull/17474

Requesting from @megglos review as he created also previously the github release workflow (migrated from jenkins). Happy for your input.

Most recent run https://github.com/camunda/zeebe/actions/runs/8732895298/job/23960606631 (maven took ~13 min)
Previous run https://github.com/camunda/zeebe/actions/runs/8728270279/job/23947651262 (maven took ~9 min)

Looks like depends a lot of downloading the whole internet (maven thing) ;)

Zelldon commented:
I started a new release dry run to verify whether we no longer see OOM's from gpg.

https://github.com/camunda/zeebe/actions/runs/8702266733/job/23865956642

Zelldon commented:
```
Error: ROR] The build could not read 2 projects -> [Help 1]
Error: ROR]   
Error: ROR]   The project io.camunda:zeebe-qa:8.6.0-SNAPSHOT (/runner/_work/zeebe/zeebe/target/checkout/zeebe/qa/pom.xml) has 1 error
Error: ROR]     Non-resolvable parent POM for io.camunda:zeebe-qa:8.6.0-SNAPSHOT: The following artifacts could not be resolved: io.camunda:zeebe-parent:pom:8.6.0-SNAPSHOT (absent): Could not find artifact io.camunda:zeebe-parent:pom:8.6.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 6, column 11 -> [Help 2]
Error: ROR]   
Error: ROR]   The project io.camunda:operate-qa:8.6.0-SNAPSHOT (/runner/_work/zeebe/zeebe/target/checkout/operate/qa/pom.xml) has 1 error
Error: ROR]     Non-resolvable parent POM for io.camunda:operate-qa:8.6.0-SNAPSHOT: The following artifacts could not be resolved: io.camunda:operate-parent:pom:8.6.0-SNAPSHOT (absent): Could not find artifact io.camunda:operate-parent:pom:8.6.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 5, column 11 -> [Help 2]
Error: ROR] 
Error: ROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
Error: ROR] Re-run Maven using the -X switch to enable full debug logging.
Error: ROR] 
Error: ROR] For more information about the errors and possible solutions, please read the following articles:
Error: ROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
Error: ROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
```

Looks like we have some issues with this approach :/

Zelldon commented:
@romansmirnov I don't think that I get if fixed today. Not sure what goes wrong with the release process when we not include the modules (🤔) I tried several things now, but still https://github.com/camunda/zeebe/actions/runs/8702266733

We might want to merge first https://github.com/camunda/zeebe/pull/17474 and enable the dry runs again and then check how we can improve from there.

Zelldon commented:
Ok looks like the previous run, was using an old version of the branch (looks like retry doesn't work as I expected to)

Here is a more recent run https://github.com/camunda/zeebe/actions/runs/8718431223/job/23915661309 where it was working.

@romansmirnov I'm still not 100% sure whether it is ok to do it like this, as maybe the versions are then not updated for QA modules 🤔 Wdyt?

Zelldon commented:
With the recent commit I also apply again spotless (I thought at the begin this is the problem) it seem to work as well https://github.com/camunda/zeebe/actions/runs/8718985223

I like the current approach more, but still unsure whether it makes a problem. I might have to try it locally how the files differ (whether versions have been updated, etc.)

Zelldon commented:
Executing the following locally (for reference):

```
export RELEASE_VERSION=0.8.2;
export DEVELOPMENT_VERSION=0.0.0-SNAPSHOT;
export PUSH_CHANGES=false;
export SKIP_REPO_DEPLOY=true;
export ZBCTL_ROOT_DIR=$(pwd);
./mvnw release:prepare release:perform -B \
    -Dgpg.passphrase="moep" \
    -Dresume=false \
    -Dtag=${RELEASE_VERSION} \
    -DreleaseVersion=${RELEASE_VERSION} \
    -DdevelopmentVersion=${DEVELOPMENT_VERSION} \
    -DpushChanges=${PUSH_CHANGES} \
    -DremoteTagging=${PUSH_CHANGES} \
    -DlocalCheckout=true \
    -DcompletionGoals="spotless:apply" \
    -P-integration-tests,-autoFormat \
    -Darguments='-T1C -P-integration-tests,-autoFormat -DskipChecks=true -DskipTests=true -Dspotless.apply.skip=false -Dskip.central.release=${SKIP_REPO_DEPLOY} -Dskip.camunda.release=${SKIP_REPO_DEPLOY} -Dzbctl.force -Dzbctl.rootDir=${ZBCTL_ROOT_DIR} -Dgpg.passphrase="***"'
```

Zelldon commented:
As expected QA is untouched:

```
On branch ck-release-ignore-qa
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   bom/pom.xml
	modified:   build-tools/pom.xml
	modified:   dist/pom.xml
	modified:   distro/pom.xml
	modified:   operate/archiver/pom.xml
	modified:   operate/client/package.json
	modified:   operate/client/pom.xml
	modified:   operate/common/pom.xml
	modified:   operate/data-generator/pom.xml
	modified:   operate/importer-8_5/pom.xml
	modified:   operate/importer-8_6/pom.xml
	modified:   operate/importer-common/pom.xml
	modified:   operate/importer/pom.xml
	modified:   operate/mvc-auth-commons/pom.xml
	modified:   operate/pom.xml
	modified:   operate/schema/pom.xml
	modified:   operate/webapp/pom.xml
	modified:   parent/pom.xml
	modified:   pom.xml
	modified:   spring-boot-starter-camunda-sdk/pom.xml
	modified:   zeebe/atomix/cluster/pom.xml
	modified:   zeebe/atomix/pom.xml
	modified:   zeebe/atomix/utils/pom.xml
	modified:   zeebe/auth/pom.xml
	modified:   zeebe/backup-stores/azure/pom.xml
	modified:   zeebe/backup-stores/common/pom.xml
	modified:   zeebe/backup-stores/gcs/pom.xml
	modified:   zeebe/backup-stores/s3/pom.xml
	modified:   zeebe/backup-stores/testkit/pom.xml
	modified:   zeebe/backup/pom.xml
	modified:   zeebe/benchmarks/project/pom.xml
	modified:   zeebe/bpmn-model/pom.xml
	modified:   zeebe/broker-client/pom.xml
	modified:   zeebe/broker/pom.xml
	modified:   zeebe/clients/java/pom.xml
	modified:   zeebe/dmn/pom.xml
	modified:   zeebe/engine/pom.xml
	modified:   zeebe/exporter-api/pom.xml
	modified:   zeebe/exporter-test/pom.xml
	modified:   zeebe/exporters/elasticsearch-exporter/pom.xml
	modified:   zeebe/exporters/opensearch-exporter/pom.xml
	modified:   zeebe/expression-language/pom.xml
	modified:   zeebe/feel/pom.xml
	modified:   zeebe/gateway-protocol-impl/pom.xml
	modified:   zeebe/gateway-protocol/pom.xml
	modified:   zeebe/gateway-rest/pom.xml
	modified:   zeebe/gateway/pom.xml
	modified:   zeebe/journal/pom.xml
	modified:   zeebe/logstreams/pom.xml
	modified:   zeebe/msgpack-core/pom.xml
	modified:   zeebe/msgpack-value/pom.xml
	modified:   zeebe/pom.xml
	modified:   zeebe/protocol-asserts/pom.xml
	modified:   zeebe/protocol-impl/pom.xml
	modified:   zeebe/protocol-jackson/pom.xml
	modified:   zeebe/protocol-test-util/pom.xml
	modified:   zeebe/protocol/pom.xml
	modified:   zeebe/restore/pom.xml
	modified:   zeebe/scheduler/pom.xml
	modified:   zeebe/snapshot/pom.xml
	modified:   zeebe/stream-platform/pom.xml
	modified:   zeebe/test-util/pom.xml
	modified:   zeebe/topology/pom.xml
	modified:   zeebe/transport/pom.xml
	modified:   zeebe/util/pom.xml
	modified:   zeebe/zb-db/pom.xml
```

Meaning the versions are not updated for QA after running the release 🤔 we might be able to split the prepare release and do the actual release. 🤔 💭 

But would be open for more input. @tmetzke @romansmirnov do you know/remember how it is done in C7? I can see that we [release different distros etc](https://github.com/camunda/camunda-bpm-release-scripts/blob/master/build-mvn-poms.properties#L16), but not sure how does this work with the release plugin.

Maybe @megglos or @oleschoenburg has also some idea, to skip the QA/IT on release.

Zelldon commented:
Ok feels like C7 does this manually ? 🤔 https://github.com/camunda/camunda-bpm-release-scripts/blob/master/build-mvn-poms.xml#L55-L80

tmetzke commented:
If I remember correctly, C7 simply doesn't use the release plugin. The release is built manually, adjusting all versions in all modules, deploying only specific modules, and adjusting versions in all modules afterward again.

tmetzke commented:
Can we adjust the deploy or staging plugins for the qa modules so they are considered in the release but not deployed publicly?

Zelldon commented:
> Can we adjust the deploy or staging plugins for the qa modules so they are considered in the release but not deployed publicly?

Yeah we might be able to skip the deploy phase. I will try this next if my last commit doesn't work where I split the perform and release steps.

Zelldon commented:
I reworked the release workflow more, got it finally working and even the build time down from 15-20 to 9 minutes

https://github.com/camunda/zeebe/actions/runs/8728270279/job/23947651262

I will clean up the PR asap, and ping for a review.

Zelldon commented:
I reworked the PR (cleaned up the commits, adjusted the description) and asked Meggle for a review. 

With the recent changes we reduce the build time by half.

megglos commented on operate/pom.xml:
```diff
@@ -257,6 +256,19 @@
+    <profile>
+      <id>integration-tests</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
```
💭  we already have the `skipTests` property can't we use that for ITs too? then you wouldn't need that profile and have one property to disable all sorts of tests?

Zelldon replied:
No because this only skips tests but still compiles and generates gpg and all these sort of stuff.

tmetzke replied:
Yeah, we really want to completely exclude the whole module when releasing, so profiles usually are the best way to go. Or skipping the plugins in those modules that you want to avoid running (e.g. gpg signing, deploying, etc.). Depends a bit on much you want to skip. Usually, it's clearer and cleaner using profiles like this.

tmetzke replied:
> 💭 I think it's fine here generally. However, as soon as you enable ANY other profile, this won't trigger anymore. That is a typical "what the heck is going on?" moment with Maven 😄

Expanding on this, this even happens when specific os- or JDK version-specific profiles are activated. Those `activeByDefault` profiles don't run by default anymore then. 😄

tmetzke replied:
We could use a rather hacky activation that always runs the QA as well unless you explicitly exclude it:
```suggestion
        <file>
          <exists>pom.xml</exists>
        </file>
```

oleschoenburg replied:
Is `-Dmaven.test.skip=true` an option? that should skip compilation altogether.

Zelldon replied:
That is interesting this would at least also avoid for the other modules. 👍

 But my point for the profile is to really do nothing with these modules as you would still verify and generate sources for these, right?

But I'm also fine to remove the profile as due to the other refactoring I did later it is maybe not that much of an impact anymore (just unnecessary work we do)

Zelldon replied:
@oleschoenburg I tried this shortly via `-Dmaven.test.skip=true` https://github.com/camunda/zeebe/pull/17497/commits/0b9a1c3eef3f12d73f16e2aeb493999f80244253

I ran into the [same issue](https://github.com/camunda/zeebe/actions/runs/8774388988/job/24075524714#step:14:10597) when I tried to compile only the project during prepare (I tested several approaches earlier). The issue is that the [snapshot project depends on a testjar that is not compiled if we use this flag](https://github.com/camunda/zeebe/blob/main/zeebe/snapshot/pom.xml#L65-L71) (or use only compile), that is why I had to do this with the profile. 

I see it as the only option right now, and would like to go ahead with it. If you see a different option please feel free to improve it afterwards.

megglos commented on operate/pom.xml:
```diff
@@ -257,6 +256,19 @@
     </profile>
-  </profiles>
 
+    <!-- 
+      Integration tests are run per default, but having it in a profile allows us to exclude it easily.
+      That is useful for example for creating releases.
+       -->
```
🔧 Following up on the discussion I would suggest the following to make clear we want to ignore the whole module not just skip test runs.
```suggestion
    <!-- 
      The QA module is enabled per default, but having it in a profile allows us to exclude it easily.
      That is useful for example for creating releases.
       -->
```

megglos commented on zeebe/pom.xml:
```diff
@@ -61,4 +60,20 @@
   </modules>
+
+  <profiles>
+    <!-- 
```
🔧 Following up on the [discussion](https://github.com/camunda/zeebe/pull/17497/files#r1570100625) I would suggest the following to make clear we want to ignore the whole module not just skip test runs.
```suggestion
    <!-- 
      The QA module is enabled per default, but having it in a profile allows us to exclude it easily.
      That is useful for example for creating releases.
       -->
```

megglos reviewed (approved):
🚀

Zelldon commented on zeebe/pom.xml:
```diff
@@ -61,4 +60,20 @@
+
+  <profiles>
+    <!-- 
+      Integration tests are run per default, but having it in a profile allows us to exclude it easily.
+      That is useful for example for creating releases.
```
```suggestion
      The QA module is enabled by default, but having it in a profile allows us to exclude it easily.
      That is useful for example for creating releases.
```
